src/osd/OSD.h

```diff
@@ -539,52 +539,76 @@ class OSDService {
    * working from old maps.
    */
 private:
+  RWLock next_osdmap_lock = {"OSDService::next_osdmap_lock"}; // orders before publish
+  Mutex next_osdmap_wait_lock = {"OSDService::next_osdmap_wait_lock"};
+  Cond next_osdmap_wait_cond;
   OSDMapRef next_osdmap;
-  Cond pre_publish_cond;
+
+  /// map epochs reserved below
+  map<epoch_t, std::atomic_int> map_reservations;
+
+  /// pointer to next_osdmap's reservation slot
+  map<epoch_t,std::atomic_int>::iterator next_map_reservation;
+
 
 public:
   void pre_publish_map(OSDMapRef map) {
-    Mutex::Locker l(pre_publish_lock);
+    RWLock::WLocker l(next_osdmap_lock);
     next_osdmap = std::move(map);
+    map_reservations[next_osdmap->get_epoch()];
+    next_map_reservation = map_reservations.find(next_osdmap->get_epoch());
   }
 
   void activate_map();
-  /// map epochs reserved below
-  map<epoch_t, unsigned> map_reservations;
 
   /// gets ref to next_osdmap and registers the epoch as reserved
   OSDMapRef get_nextmap_reserved() {
-    Mutex::Locker l(pre_publish_lock);
+    RWLock::RLocker l(next_osdmap_lock);
     if (!next_osdmap)
       return OSDMapRef();
-    epoch_t e = next_osdmap->get_epoch();
-    map<epoch_t, unsigned>::iterator i =
-      map_reservations.insert(make_pair(e, 0)).first;
-    i->second++;
+    ++next_map_reservation->second;
     return next_osdmap;
   }
+
   /// releases reservation on map
   void release_map(OSDMapRef osdmap) {
-    Mutex::Locker l(pre_publish_lock);
-    map<epoch_t, unsigned>::iterator i =
-      map_reservations.find(osdmap->get_epoch());
-    assert(i != map_reservations.end());
-    assert(i->second > 0);
-    if (--(i->second) == 0) {
-      map_reservations.erase(i);
+    RWLock::RLocker l(next_osdmap_lock);
+    if (osdmap->get_epoch() == next_map_reservation->first) {
+      // fast path: still the latest
+      --next_map_reservation->second;
+    } else {
+      auto i = map_reservations.find(osdmap->get_epoch());
+      if (--i->second == 0) {
+	Mutex::Locker l(next_osdmap_wait_lock);
+	next_osdmap_wait_cond.Signal();
+      }
     }
-    pre_publish_cond.Signal();
   }
+
   /// blocks until there are no reserved maps prior to next_osdmap
   void await_reserved_maps() {
-    Mutex::Locker l(pre_publish_lock);
-    assert(next_osdmap);
     while (true) {
-      map<epoch_t, unsigned>::const_iterator i = map_reservations.cbegin();
-      if (i == map_reservations.cend() || i->first >= next_osdmap->get_epoch()) {
+      bool wait = false;
+      {
+	RWLock::WLocker l(next_osdmap_lock);
+	auto i = map_reservations.begin();
+	while (i != map_reservations.end()) {
+	  if (i->first < next_osdmap->get_epoch()) {
+	    if (i->second.load() == 0) {
+	      // erase old, empty slot
+	      i = map_reservations.erase(i);
+	      continue;
+	    }
+	    wait = true;
+	  }
+	  break;
+	}
+      }
```

Comment:
if wait is true and this thread release next_osdmap_lock, another thread call release_map before this thread entering "Wait", it will cause this thread Wait infinitely